packages/api-contract/src/MetaRegistry.ts

```diff
@@ -0,0 +1,298 @@
+    if (def === 'builtin') {
+      if (!id) {
+        throw new Error('Invalid builtin type metadata found');
+      }
+
+      // builtin primitive type
+      if (typeof id === 'string') {
+        return {
+          info: t.TypeDefInfo.Plain,
+          type: id
+        };
+      }
+
+      // builtin tuple
+      if (Array.isArray(id)) {
+        const sub = id.map((tupleTypeIndex: number): t.TypeDef => this.typeDefFromMetaTypeAt(tupleTypeIndex));
+        return {
+          info: t.TypeDefInfo.Tuple,
+          type: encodeTuple(sub),
+          sub
+        };
+      }
+
+      // builtin array
+      const { 'array.type': vecTypeIndex, 'array.len': vecLength } = id as t.MetaTypeIdArray;
+      if (vecTypeIndex) {
+        assert(!vecLength || vecLength <= 256, `MetaRegistry: Only support for [Type; <length>], where length <= 256`);
+        assert(!typeIndex || vecTypeIndex !== typeIndex, `MetaRegistry: self-referencing registry type at index ${typeIndex}`);
+        const sub = this.typeDefFromMetaTypeAt(vecTypeIndex);
+
+        return {
+          ...(
+            vecLength
+              ? {
+                info: t.TypeDefInfo.VecFixed,
+                ext: { length: vecLength, type: '' },
+                type: `[${sub.name || sub.type};${vecLength}]`
+              }
+              : {
+                info: t.TypeDefInfo.Vec,
+                type: `Vec<${sub.name || sub.type}>`
+              }
+          ),
+          sub: this.typeDefFromMetaTypeAt(vecTypeIndex)
+        };
+      }
+    }
+
+    // enum
+    const { 'enum.variants': enumVariants } = def as t.MetaTypeDefEnum;
+    if (enumVariants) {
+      const sub = enumVariants.map(
+        (variant): t.TypeDef => {
+          const { 'unit_variant.name': unitNameIndex } = variant as t.MetaTypeDefEnumVariantUnit;
+          if (unitNameIndex) {
+            return {
+              info: t.TypeDefInfo.Plain,
+              name: this.stringAt(unitNameIndex),
+              type: 'Null'
+            };
+          }
+
+          const { 'tuple_struct_variant.name': tupleStructVariantNameIndex } = variant as t.MetaTypeDefEnumVariantTupleStruct;
+          if (tupleStructVariantNameIndex) {
+            return this.typeDefForTupleStruct(variant as t.MetaTypeDefEnumVariantTupleStruct);
+          }
+
+          const { 'struct_variant.name': structVariantNameIndex } = variant as t.MetaTypeDefEnumVariantStruct;
+          if (structVariantNameIndex) {
+            return this.typeDefForStruct(variant as t.MetaTypeDefEnumVariantStruct);
+          }
+
+          return {
+            info: t.TypeDefInfo.Null,
+            type: 'Null'
+          };
+        }
+      );
+
+      return {
+        info: t.TypeDefInfo.Enum,
+        type: encodeEnum(sub),
+        sub
+      };
+    }
+
+    // clike enum
+    const { 'clike_enum.variants': clikeEnumVariants } = def as t.MetaTypeDefClikeEnum;
```

Comment:
(nit) as above

Reply:
At this point this has turned into a real absolute monster function. Can we clean it up to be in manageable chunks, would rather not come back at some point to make CodeClimate happy